// simple-node-server/dist/lib/container/default.logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const colors = require("colors");
const SimpleLogger = require("simple-node-logger");
class Logger {
    static initialize(logFilePath, timestampFormat = 'YYYY-MM-DD HH:mm:ss.SSS') {
        Logger.logger = SimpleLogger.createSimpleLogger({
            logFilePath: logFilePath,
            timestampFormat: timestampFormat
        });
    }
    static info(logable) {
        if (logable) {
            if (typeof logable === 'string') {
                Logger.logger.info(logable);
            }
            else {
                const err = logable;
                Logger.logger.info(colors.blue(`${err.name}\n, ${err.message}\n, ${err.stack}`));
            }
        }
    }
    static warning(logable) {
        if (logable) {
            if (typeof logable === 'string') {
                Logger.logger.warning(colors.black.bgYellow(logable));
            }
            else {
                const err = logable;
                Logger.logger.warning(colors.yellow(`${err.name}\n, ${err.message}\n, ${err.stack}`));
            }
        }
    }
    static error(logable) {
        if (logable) {
            if (typeof logable === 'string') {
                Logger.logger.error(colors.white.bgRed(logable));
            }
            else {
                const err = logable;
                Logger.logger.error(colors.red(`${err.name}\n, ${err.message}\n, ${err.stack}`));
            }
        }
    }
}
Logger.logger = SimpleLogger.createSimpleLogger({
    logFilePath: 'log/logfile.log',
    timestampFormat: 'YYYY-MM-DD HH:mm:ss.SSS'
});
exports.default = Logger;
