// simple-node-server/dist/lib/classes/cookie.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class CookieParser {
    constructor(cookieString) {
        this.parsedCookie = {};
        this.cookieString = cookieString;
        this.parse();
    }
    parse() {
        const array = this.cookieString.split(';');
        array.forEach((data) => {
            this.extract(data);
        });
    }
    extract(data) {
        const keyValue = data.split('=');
        switch (keyValue[0]) {
            case 'Expires':
                this.parsedCookie.expires = new Date(keyValue[1]).getTime();
                break;
            case 'Secure':
                this.parsedCookie.secure = true;
                break;
            case 'HttpOnly':
                this.parsedCookie.httpOnly = true;
                break;
            default:
                this.parsedCookie.name = keyValue[0];
                this.parsedCookie.value = keyValue[1];
                break;
        }
    }
    getParsedCookie() {
        return this.parsedCookie;
    }
}
exports.CookieParser = CookieParser;
